imports/plugins/core/orders/server/no-meteor/mutations/createRefund.js

```diff
@@ -0,0 +1,114 @@
+  const { Orders } = collections;
+
+  // First verify that this order actually exists
+  const order = await Orders.findOne({ _id: orderId });
+  if (!order) throw new ReactionError("not-found", "Order not found");
+
+  // Allow refund for accounts with orders or "order/fulfillment" permissions.
+  // When called internally by another plugin, context.isInternalCall
+  // can be set to `true` to disable this check.
+  if (!isInternalCall && !userHasPermission(["orders", "order/fulfillment"], order.shopId)) {
+    throw new ReactionError("access-denied", "Access Denied");
+  }
+
+  // Verify payment exists
+  const payment = (order.payments || []).find((pmt) => pmt._id === paymentId);
+  if (!payment) throw new ReactionError("not-found", "Payment not found");
+
+  // Verify that payment method will accept refunds
+  const { name } = payment;
+  const { canRefund, functions } = getPaymentMethodConfigByName(name);
+  if (!canRefund) throw new ReactionError("invalid", "Refunding not supported");
+
+  // Find total of any previous refunds, and determine how much balance is left
+  // on this payment that can be refunded.
+  const previousRefunds = await functions.listRefunds(context, payment);
+  const previousRefundsTotal = previousRefunds.reduce((sum, refund) => sum + refund.amount, 0);
+  const amountNotYetRefunded = payment.amount - previousRefundsTotal;
+
+  // Refund the lower amount between `amountNotYetRefunded`, and the `amount` that was passed
+  const amountToRefund = Math.min(amountNotYetRefunded, amount);
+  const result = await functions.createRefund(context, payment, amountToRefund, reason);
+
+  if (!result.saved) {
+    Logger.fatal("Attempt to refund payment failed", order._id, paymentId, result.error);
+    throw new ReactionError("Attempt to refund payment failed", result.error);
+  }
+
+  // List refunds to see if we've now refunded the full amount of this payment
+  const allRefunds = await functions.listRefunds(context, payment);
+  const newRefundsTotal = allRefunds.reduce((sum, refund) => sum + refund.amount, 0);
+
+  // There could be JS math errors so we round to 3 decimal places when comparing
+  const isFullyRefunded = accounting.toFixed(newRefundsTotal, 3) === accounting.toFixed(payment.amount, 3);
+
+
+  const { modifiedCount, value: updatedOrder } = await Orders.findOneAndUpdate(
+    {
+      "_id": orderId,
+      "payments._id": paymentId
+    },
+    {
+      $set: {
+        "payments.$.status": isFullyRefunded ? "refunded" : "partialRefund"
+      },
+      $push: {
+        "payments.$.transactions": result
+      }
+    }
+  );
+
+  if (modifiedCount === 0 || !updatedOrder) throw new ReactionError("server-error", "Unable to update order");
+
+  await appEvents.emit("afterOrderUpdate", {
+    order: updatedOrder,
+    updatedBy: userId
+  });
+
+  // Send email to notify customer of a refund
+  sendOrderEmail(context, order, "refunded");
```

Comment:
I think maybe we should pass `updatedOrder` to this function. The old code passed `order` too but seems maybe a bug.